# Tutoriais de Primeiros Passos no Pandas

## 1- Que Tipo de Dados o Pandas Trata?

In [1]:
# Importando o pacote
import pandas as pd

### Representação da tabela de dados/DataFrame do Pandas 

Um **DataFrame** é uma **estrutura de dados bidimensional** que pode armazenar dados de diferentes tipos (incluindo caracteres, inteiros, valores de ponto flutuante, dados categóricos e muito mais) em colunas. É semelhante a uma planilha do Excel ou uma tabela SQL.

In [2]:
# Armazenando dados manualmente em um DataFrame
df = pd.DataFrame(
    {
        "Name": [
            "Braund, Mr. Owen Harris",
            "Allen, Mr. William Henry",
            "Bonnell, Miss. Elizabeth",
        ],
        "Age": [22, 35, 58],
        "Sex": ["male", "male", "female"],
    }
)

df

Name  Age     Sex
0   Braund, Mr. Owen Harris   22    male
1  Allen, Mr. William Henry   35    male
2  Bonnell, Miss. Elizabeth   58  female

### Cada coluna em um DataFrame é uma Series

Uma **Series** não possui rótulos de coluna, pois tem apenas uma única coluna. No entanto, possui rótulos de linha.

In [3]:
# Criando uma Series
ages = pd.Series([22, 35, 58], name="Age")

In [4]:
# Ao selecionar uma única coluna de um DataFrame, o resultado é uma Series
df["Age"]

0    22
1    35
2    58
Name: Age, dtype: int64

### Fazendo algo com um DataFrame ou Series 


In [5]:
# Retorna o valor mais alto para a coluna selecionada
df["Age"].max()

58

In [6]:
# Fornece algumas estatísticas básicas dos dados numéricos de um DataFrame, por padrão dados textuais não são considerados
df.describe()

Age
count   3.000000
mean   38.333333
std    18.230012
min    22.000000
25%    28.500000
50%    35.000000
75%    46.500000
max    58.000000

## 2- Como Ler e Escrever Dados Tabulares?

O Pandas suporta muitos formatos de arquivo ou fontes de dados diferentes prontos para uso (csv, excel, sql, json, parquet, ...), cada um deles com o prefixo `read_*`.
\
\
**IMPORTANTE**: Para o conjunto de dados do Titanic, armazenado em CSV. Os dados consistem nas seguintes colunas de dados:

- **PassengerId**: ID do passageiro
- **Survived**: indica se o passageiro sobreviveu. `0` para sim e `1` para não
- **Pclass**: uma das 3 classes de ingresso: classe `1`, classe `2` e classe `3`
- **Name**: nome do passageiro
- **Sex**: gênero do passageiro
- **Age**: idade do passageiro
- **SibSp**: número de irmãos ou cônjuges a bordo
- **Parch**: número de pais ou filhos a bordo
- **Ticket**: número do bilhete do passageiro
- **Fare**: indica a tarifa
- **Cabin**: número da cabine do passageiro
- **Embarked**: porto de embarque

In [7]:
titanic = pd.read_csv("datasets/titanic.csv")
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [8]:
# Exibe as primeiras N linhas de um DataFrame, para as últimas N linhas use o método tail()
titanic.head(8)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S

In [9]:
# Informa como o Pandas interpretou cada um dos tipos de dados da coluna 
titanic.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [10]:
# Armazena os dados como um arquivo Excel. O parâmetro index=False, indica que os rótulos do índice de linha não devem ser salvos na planilha
titanic.to_excel("titanic.xlsx", sheet_name="passengers", index=False)

In [11]:
# Fornece informações técnicas sobre um DataFrame
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Podemos interpretar sua saída da seguinte forma:

- É um DataFrame
- Existem 891 entradas, ou seja, 891 linhas
- Cada linha possui um rótulo de linha (também conhecido como index) com valores que variam de 0 a 890
- A tabela possui 12 colunas. Algumas colunas possuem valores ausentes, são elas: Age, Cabin e Embarked
- As colunas possuem diferentes tipos de dados
- Os tipos de dados nas diferentes colunas são resumidos em dtypes
- A quantidade aproximada de RAM usada para armazenar o DataFrame é de 83.7 KB

## 3- Selecionando um Subconjunto de um DataFrame

### Selecionando colunas específicas de um DataFrame

In [12]:
ages = titanic["Age"]
ages.head()

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
Name: Age, dtype: float64

In [13]:
# Cada coluna em um DataFrame é uma Series
type(titanic["Age"])

pandas.core.series.Series

In [14]:
# O atributo shape contêm o número de linhas e colunas: (nrows, ncolumns). Uma Series é unidimensional e apenas o número de linhas é retornado
titanic["Age"].shape

(891,)

In [15]:
# Para selecionar várias colunas, use uma lista Python
age_sex = titanic[["Age", "Sex"]]
age_sex.head()

Age     Sex
0  22.0    male
1  38.0  female
2  26.0  female
3  35.0  female
4  35.0    male

In [16]:
# O tipo de dados retornado é um DataFrame
type(titanic[["Age", "Sex"]])

pandas.core.frame.DataFrame

In [17]:
#  Um DataFrame é bidimensional e o número de linhas e colunas é retornado
titanic[["Age", "Sex"]].shape

(891, 2)

### Filtrando linhas específicas de um DataFrame 

In [18]:
# A condição dentro dos colchetes de seleção verifica em quais linhas a coluna Age tem valor maior do que 35
above_35 = titanic[titanic["Age"] > 35]
above_35.head()

PassengerId  Survived  Pclass  \
1             2         1       1   
6             7         0       1   
11           12         1       1   
13           14         0       3   
15           16         1       2   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   

    Parch    Ticket     Fare Cabin Embarked  
1       0  PC 17599  71.2833   C85        C  
6       0     17463  51.8625   E46        S  
11      0    113783  26.5500  C103        S  
13      5    347082  31.2750   NaN        S  
15      0    248706  16.0000   NaN        S

A saída da expressão condicional é uma Series de valores booleanos (`True` ou `False`) com o mesmo número de linhas que o DataFrame original. Esses valores booleanos podem ser usados para filtrar o DataFrame e somente as linhas para as quais o valor for `True` serão retornadas.

In [19]:
titanic["Age"] > 35

0      False
1       True
2      False
3      False
4      False
       ...  
886    False
887    False
888    False
889    False
890    False
Name: Age, Length: 891, dtype: bool

In [20]:
# isin() retorna True para cada linha em que os valores estão na lista fornecida
class_23 = titanic[titanic["Pclass"].isin([2, 3])]
class_23.head()

PassengerId  Survived  Pclass                            Name     Sex  \
0            1         0       3         Braund, Mr. Owen Harris    male   
2            3         1       3          Heikkinen, Miss. Laina  female   
4            5         0       3        Allen, Mr. William Henry    male   
5            6         0       3                Moran, Mr. James    male   
7            8         0       3  Palsson, Master. Gosta Leonard    male   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0  22.0      1      0         A/5 21171   7.2500   NaN        S  
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4  35.0      0      0            373450   8.0500   NaN        S  
5   NaN      0      0            330877   8.4583   NaN        Q  
7   2.0      3      1            349909  21.0750   NaN        S

**Observação**: Ao combinar múltiplas instruções condicionais, cada condição deve estar entre parênteses. Além disso, você deve usar os operadores lógicos `|` e `&`.

In [21]:
# Semelhante a isin([2, 3])
class_23 = titanic[(titanic["Pclass"] == 2) | (titanic["Pclass"] == 3)]
class_23.head()

PassengerId  Survived  Pclass                            Name     Sex  \
0            1         0       3         Braund, Mr. Owen Harris    male   
2            3         1       3          Heikkinen, Miss. Laina  female   
4            5         0       3        Allen, Mr. William Henry    male   
5            6         0       3                Moran, Mr. James    male   
7            8         0       3  Palsson, Master. Gosta Leonard    male   

    Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
0  22.0      1      0         A/5 21171   7.2500   NaN        S  
2  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S  
4  35.0      0      0            373450   8.0500   NaN        S  
5   NaN      0      0            330877   8.4583   NaN        Q  
7   2.0      3      1            349909  21.0750   NaN        S

In [22]:
# Retorna True para cada linha onde os valores são existentes (não ausentes)
age_no_na = titanic[titanic["Age"].notna()]
age_no_na.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### Selecionando linhas e colunas específicas de um DataFrame

Nesse caso, um subconjunto de linhas e colunas é criado de uma só vez e apenas usar colchetes de seleção `[]` não é mais suficiente. Os operadores `loc`/`iloc` são obrigatórios antes dos colchetes de seleção. Ao usar `loc`/`iloc`, a parte antes da vírgula são as linhas e a parte depois da vírgula são as colunas que você deseja selecionar.

In [23]:
# Retorna nomes dos passageiros com mais de 35 anos
adult_names = titanic.loc[titanic["Age"] > 35, "Name"]
adult_names.head()

1     Cumings, Mrs. John Bradley (Florence Briggs Th...
6                               McCarthy, Mr. Timothy J
11                             Bonnell, Miss. Elizabeth
13                          Andersson, Mr. Anders Johan
15                     Hewlett, Mrs. (Mary D Kingcome) 
Name: Name, dtype: object

Ao usar nomes de colunas, rótulos de linhas ou uma expressão de condição, use o operador `loc`. Tanto para a parte antes quanto depois da vírgula, você pode usar um único rótulo, uma lista de rótulos, uma fatia de rótulos, uma expressão condicional ou dois pontos. Usar dois pontos especifica que você deseja selecionar todas as linhas ou colunas.

In [24]:
# Retorna as linhas 10 a 24 e as colunas 3 a 4
titanic.iloc[9:25, 2:5]

Pclass                                               Name     Sex
9        2                Nasser, Mrs. Nicholas (Adele Achem)  female
10       3                    Sandstrom, Miss. Marguerite Rut  female
11       1                           Bonnell, Miss. Elizabeth  female
12       3                     Saundercock, Mr. William Henry    male
13       3                        Andersson, Mr. Anders Johan    male
14       3               Vestrom, Miss. Hulda Amanda Adolfina  female
15       2                   Hewlett, Mrs. (Mary D Kingcome)   female
16       3                               Rice, Master. Eugene    male
17       2                       Williams, Mr. Charles Eugene    male
18       3  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female
19       3                            Masselmani, Mrs. Fatima  female
20       2                               Fynney, Mr. Joseph J    male
21       2                              Beesley, Mr. Lawrence    male
22       3                        McGowan, Miss. Anna "Annie"  female
23       1                       Sloper, Mr. William Thompson    male
24       3                      Palsson, Miss. Torborg Danira  female

Ao selecionar linhas e/ou colunas específicas com `loc`/`iloc`, novos valores podem ser atribuídos aos dados selecionados.

In [25]:
# Atribui o nome anonymous aos primeiros 3 elementos da quarta coluna
titanic.iloc[0:3, 3] = "anonymous"
titanic.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                           Name     Sex   Age  SibSp  Parch  \
0                                     anonymous    male  22.0      1      0   
1                                     anonymous  female  38.0      1      0   
2                                     anonymous  female  26.0      0      0   
3  Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1      0   
4                      Allen, Mr. William Henry    male  35.0      0      0   

             Ticket     Fare Cabin Embarked  
0         A/5 21171   7.2500   NaN        S  
1          PC 17599  71.2833   C85        C  
2  STON/O2. 3101282   7.9250   NaN        S  
3            113803  53.1000  C123        S  
4            373450   8.0500   NaN        S

**LEMBRE-SE**
`loc`: seleciona linhas e/ou colunas específicas usando os nomes de linhas e colunas
\
`iloc`: seleciona linhas e/ou colunas específicas usando as posições na tabela